drivers/vulkan/vulkan_context.cpp

```diff
@@ -2291,10 +2305,6 @@ Error VulkanContext::prepare_buffers() {
 
 	VkResult err;
 
-	// Ensure no more than FRAME_LAG renderings are outstanding.
-	vkWaitForFences(device, 1, &fences[frame_index], VK_TRUE, UINT64_MAX);
```

Comment:
Would the wait still be OK to happen here if it was waiting for the, properly rotated, `frame_index - 1`? Wouldn't that the a more CPU-efficient, given that between `swap_buffers` and the next `prepare_buffers` the application could be using the time for something else than waiting? (Disclaimer: rather than assuming you're wrong in your changes, this is the way for me to understand them.)

Reply:
If what you're asking is (ignoring performance) if we could wait here doing:

`vkWaitForFences(device, 1, &fences[(frame_index + 1u) % frame_count], VK_TRUE, UINT64_MAX);`

Then the answer is yes, that would be functionally equivalent.

Now if we consider performance, you're asking the comparison:

```cpp
// What you propose
wait( A );
submit( B );

// vs 

// What I wrote
submit( B );
wait( A );
```

In both cases we wait for A. But in my version by the time we're done with the CPU code submission, the GPU may already be done and the wait won't happen.

In fact we should delay `wait( A )` as much as possible (in order to lower the chance we even wait at all). It's just that due to the current complexity of Godot, the current place where I've placed it is the safest spot that I could delay it.

When would we be forced to wait (i.e. cannot delay any longer)?

 - We need to destroy a vulkan handle (e.g. a framebuffer, render pass, possibly a shader & pso if they were phased out N frames ago).
 - We need to map a buffer for reading or writing.
 - We need to reuse (reset) a command buffer for preparing commands.

Reply:
So, you're doing this:
```
swap_buffers():
	submit(frame_index)
	frame_index += (frame_index + 1) % N
	wait(frame_index)
```

Can't that be changed to this?:
```
prepare_buffers():
    wait(frame_index)

swap_buffers():
	submit(frame_index)
	frame_index += (frame_index + 1) % N
```

I think stuff happens between `swap_buffers()` and the next `prepare_buffers()`. Wouldn't this allow that stuff to use the CPU in cases where it wouldn't have any choice but wait for the GPU? In other words, I think this is maybe the chance you mentioned about waiting as late as possible.

Reply:
Makes sense. I was about to suggest moving the call to `_free_pending_resources()` to a safe spot, but if there are other potential usafe operations between, I'm happy with the change. I hope we can polish that part down to perfection in the future.

Reply:
Hmm, I don't think that's what's happening here.

YES, the original order is:
``` 
	vkWaitForFences(device, 1, &fences[frame_index], VK_TRUE, UINT64_MAX);
	vkResetFences(device, 1, &fences[frame_index]);
```
but this is in prepare_buffers, which happens much later. The vkResetFences should probably go where it does in your PR (much cleaner), but the wait should happen right at the beginning of prepare_buffres.

`frame_index` is incremented after swap_buffers, meaning that each time we arrive here, we have already done all other work and are just getting ready to `prepare_screen_for_drawing`.

We're basically waiting at the last possible moment to acquire the image, just before rendering.

(technically, there is an even later time, this is before the final blit, but that's terrible for tiled renderers and mobile, and might degrade performance on desktop for slightly less micro stutter)


I'm currently examining an issue that frames actually get shown out of order, and even repeatedly. I don't know the cause yet, but I suspect keeping separate "buffers" around instead of allocating one set per swapchain image is likely part of the problem. (not a problem of this PR, but maybe you have some insights on that?)

Reply:
> but the wait should happen right at the beginning of prepare_buffers

That is the ideal, yes. But as I explained in [my comment](https://github.com/godotengine/godot/pull/80566#discussion_r1328808985) if you do that right now, you'll start deleting resources that must not be deleted without waiting first.

And even if we fixed that, Godot has no mechanism in place to prevent or detect nobody in future PRs accidentally introduces dangerous operations before the wait (that must happen after the wait).